server/block/item_frame.go

```diff
@@ -0,0 +1,166 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/entity"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// ItemFrame is a block entity that displays the item or block that is inside it.
+type ItemFrame struct {
+	empty
+	transparent
+
+	// Facing is the direction from the frame to the block.
+	Facing cube.Face
+	// Item is the item that is displayed inside the frame.
+	Item item.Stack
+	// Rotations is the number of rotations for the item in the frame. Each rotation is 45 degrees, with the exception
+	// being maps having 90 degree rotations.
+	Rotations int
+	// Drops is the chance of the item dropping when the frame is broken. In vanilla, this is always 1.0.
+	Drops float64
+	// Glowing makes the frame the glowing variant.
+	Glowing bool
+	// Large makes the frame show up large. This typically only occurs with maps.
+	Large bool
+}
+
+// Activate ...
+func (i ItemFrame) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User) bool {
+	if !i.Item.Empty() {
+		// TODO: Item frames with maps can only be rotated four times.
+		i.Rotations = (i.Rotations + 1) % 8
+		w.PlaySound(pos.Vec3Centre(), sound.ItemFrameRotate{})
+	} else if held, other := u.HeldItems(); !held.Empty() {
+		i.Item = held.Grow(-held.Count() + 1)
+		// TODO: When maps are implemented, check the item is a map, and if so, force the large boolean to true.
+		u.SetHeldItems(held.Grow(-1), other)
+		w.PlaySound(pos.Vec3Centre(), sound.ItemFrameAdd{})
+	} else {
+		return true
+	}
+
+	w.SetBlock(pos, i)
+	return true
+}
+
+// Punch ...
+func (i ItemFrame) Punch(pos cube.Pos, _ cube.Face, w *world.World, u item.User) {
+	if i.Item.Empty() {
+		return
+	}
+
+	if g, ok := u.(interface {
+		GameMode() world.GameMode
+	}); ok {
+		if rand.Float64() <= i.Drops && !g.GameMode().CreativeInventory() {
+			it := entity.NewItem(i.Item, pos.Vec3Centre())
+			it.SetVelocity(mgl64.Vec3{rand.Float64()*0.2 - 0.1, 0.2, rand.Float64()*0.2 - 0.1})
+			w.AddEntity(it)
+		}
+	}
+	i.Item, i.Rotations, i.Large = item.Stack{}, 0, false
+	w.PlaySound(pos.Vec3Centre(), sound.ItemFrameRemove{})
+	w.SetBlock(pos, i)
+}
+
+// UseOnBlock ...
+func (i ItemFrame) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) bool {
+	pos, face, used := firstReplaceable(w, pos, face, i)
+	if !used {
+		return false
+	}
+	if !w.Block(pos.Side(face.Opposite())).Model().FaceSolid(pos.Side(face.Opposite()), face, w) {
```

Comment:
This is preventing quite a lot of valid placements

Reply:
Could you give an example of a few placements that should be possible but don't work?